lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp

```diff
@@ -5013,55 +5071,226 @@ DotOpConversion::convertFMADot(triton::DotOp op, OpAdaptor adaptor,
+  int bNumPtr = 8;
+  int NK = bShape[0];
+
+  auto shapePerCTA = getShapePerCTA(dLayout);
+  auto sizePerThread = getSizePerThread(dLayout);
+
+  Value _0 = i32_val(0);
+
+  Value mContig = i32_val(sizePerThread[order[1]]);
+  Value nContig = i32_val(sizePerThread[order[0]]);
+
+  // threadId in blocked layout
+  auto threadIds = getThreadIds(thread, shapePerCTA, order, rewriter, loc);
+  Value threadIdN = threadIds[1];
+
   Value offB0 = isBRow ? mul(threadIdN, nContig) : _0;
   Value offB1 = isBRow ? _0 : mul(threadIdN, nContig);
   SmallVector<Value> bOff(bNumPtr);
   for (int i = 0; i < bNumPtr; ++i) {
     bOff[i] = add(mul(offB0, i32_val(strideB0)), mul(offB1, i32_val(strideB1)));
   }
 
-  auto aSmem = getSharedMemoryObjectFromStruct(loc, adaptor.a(), rewriter);
-  auto bSmem = getSharedMemoryObjectFromStruct(loc, adaptor.b(), rewriter);
+  auto bSmem =
+      ConvertTritonGPUOpToLLVMPatternBase::getSharedMemoryObjectFromStruct(
+          loc, llB, rewriter);
 
   Type f32PtrTy = ptr_ty(f32_ty);
-  SmallVector<Value> aPtrs(aNumPtr);
-  for (int i = 0; i < aNumPtr; ++i)
-    aPtrs[i] = gep(f32PtrTy, aSmem.base, aOff[i]);
-
   SmallVector<Value> bPtrs(bNumPtr);
   for (int i = 0; i < bNumPtr; ++i)
     bPtrs[i] = gep(f32PtrTy, bSmem.base, bOff[i]);
 
+  int N = bShape[bOrder[0]];
+  ValueTable hbs;
+
+  int nShapePerCTA = getShapePerCTAForMN(dLayout, false /*isM*/);
+  int nSizePerThread = getsizePerThreadForMN(dLayout, false /*isM*/);
+
+  for (unsigned k = 0; k < NK; ++k)
+    for (unsigned n = 0; n < N; n += nShapePerCTA)
+      for (unsigned nn = 0; nn < nSizePerThread; ++nn) {
+        Value pb = gep(f32PtrTy, bPtrs[0],
+                       i32_val((n + nn) * strideBN + k * strideBK));
+        Value vb = load(pb);
+        hbs[{n + nn, k}] = vb;
+      }
+
+  return getStructFromValueTable(hbs, rewriter, loc);
+}
+
+DotOpFMAConversionHelper::ValueTable
+DotOpFMAConversionHelper::getValueTableFromStruct(
+    Value val, int K, int n0, int shapePerCTA, int sizePerThread,
+    ConversionPatternRewriter &rewriter, Location loc) const {
+  ValueTable res;
+  auto elems = ConvertTritonGPUOpToLLVMPatternBase::getElementsFromStruct(
+      loc, val, rewriter);
+  int id = 0;
+  std::set<std::pair<int, int>> keys; // ordered
+  for (unsigned k = 0; k < K; ++k) {
+    for (unsigned m = 0; m < n0; m += shapePerCTA)
+      for (unsigned mm = 0; mm < sizePerThread; ++mm) {
+        keys.insert({m + mm, k});
+      }
+  }
+
+  for (auto &key : llvm::enumerate(keys)) {
+    res[key.value()] = elems[key.index()];
+  }
+
+  return res;
+}
+SmallVector<Value> DotOpFMAConversionHelper::getThreadIds(
+    Value threadId, ArrayRef<unsigned int> shapePerCTA,
+    ArrayRef<unsigned int> order, ConversionPatternRewriter &rewriter,
+    Location loc) const {
+  int dim = order.size();
+  SmallVector<Value> threadIds(dim);
+  for (unsigned k = 0; k < dim - 1; k++) {
+    Value dimK = i32_val(shapePerCTA[order[k]]);
+    Value rem = urem(threadId, dimK);
```

Comment:
gpu_dialect's thread_id is for getting index along one specific dimension, it is not flattened linear id.
if block size is (32, 16), this maybe not work.